### PR TITLE
[lit-html] Add classMap and styleMap directives

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (sorvell) [inline packages/lit-html/src/lib/directives/class-map.ts:100]: Why does update get the part? Users could stored it via the constructor if it's needed.

Is the idea that you can't do anything good in update without the part?

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (sorvell) [inline packages/lit-html/src/lib/directives/class-map.ts:112]: `__element` should be public or passed in separately.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (sorvell) [inline packages/lit-html/src/lib/directives/class-map.ts:118]: This is a lot of places. Need to decide in or out on `for...of`.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (sorvell) [inline packages/lit-html/src/lib/directives/class-map.ts:104]: Might be clearer to just do this block in `render`?

Comment (sorvell) [reply packages/lit-html/src/lib/directives/class-map.ts:104]: Or alternatively, and I'm not sure this is better but another take on this:
1. keep a 'isUpdate` flag and only call render if it's false.
2. do not pre-populate `previousClasses` and instead always run the rest of the code here.
3. factor the add/remove and have it only apply the change to the classList only if `isUpdate` is true.
4. return the right result based on the `isUpdate` flag.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/class-map.ts:79]: You still need a super() call

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/class-map.ts:104]: I didn't want to do this in render because it's wasted work in SSR.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/class-map.ts:112]: We'll make it public

File: packages/lit-html/src/test/directives/class-map_test.ts
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/class-map_test.ts:96]: Sorry, I forgot the behavior here. I had thought that statics could be overridden, but provided a default if they weren't in the dynamic class info. I think this is from @sorvell's initial implementation.

I guess there's two ways to look at static + dynamic classes. One is that statics provide the defaults, to be overridden by dynamic. The other is that dynamic classes are computed separately and after being determined are merged with static classes, so static classes can never be removed. That's what we have now and outside compelling reasons otherwise, I don't think we should make a breaking change here.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/class-map.ts]: btw, it's not whether `for/of` is in or out, it's whether we use them with iterables, which requires downlevel iteration. for/of with arrays is fine.

**Comment by justinfagnani** [comment]
PTAL

**Comment by justinfagnani** [comment]
I accidentally pushed styleMap to this branch, but decided to keep it since it's such a similar port. Hopefully that's ok.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/lib/directives/class-map.ts]: How does babel know if it's going to get an array or iterable, though? Doesn't it have to assume iterable?

Comment (justinfagnani) [reply packages/lit-html/src/lib/directives/class-map.ts]: It doesn't. You either turn downlevel iteration on or off. If we make sure that we don't use for/of where we know we have an iterable, then we don't force downlevel iteration on apps that compile our code.

for/of loops that take user data only need to be compiled with downlevel iteration if outside code is passing an iterable. In that case they would already need downlevel iteration.

File: packages/lit-html/src/directives/style-map.ts
Comment (kevinpschaaf) [inline packages/lit-html/src/directives/style-map.ts:61]: It's unclear why we choose to actually use the string render for first render in `classMap` but not in `styleMap`. This is another case where `render` is _only_ implemented for SSR, and otherwise unused in the client code (and as such, why the tests didn't catch the camel-casing issue in `render`).

**Review by kevinpschaaf — CHANGES_REQUESTED** [review]

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:195]: Yeah. Prettier did this in formatting

File: packages/lit-html/src/directives/style-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/style-map.ts:55]: I updated this with your SSR implementation. Also tested it two ways for comparison: One, by adding a test of render() directly. The other by returning `this.render()` from update.

**Comment by justinfagnani** [comment]
Re: classMap behavior. I filed #1278 and added a skipped test with a TODO. Let's chat about the behavior we want.

**Comment by justinfagnani** [comment]
PTAL

File: packages/lit-html/src/directives/style-map.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/directives/style-map.ts:77]: I think I added the `trim()` in when writing tests for SSR to avoid the string-based render differing from the serialization the browser does to the attribute when modifying it via the OM (this was also the reason for the space between properties). I sort of forget if there was a case where this would register as an attribute mutation, or if it was to just avoid encoding the exact serialization rules the string render did in the test. We could probably remove it for now and add it back in if necessary when we port the SSR stuff over.

File: packages/lit-html/src/directives/style-map.ts
Comment (kevinpschaaf) [reply packages/lit-html/src/directives/style-map.ts:72]: I like it!

File: packages/lit-html/src/test/directives/style-map_test.ts
Comment (kevinpschaaf) [inline packages/lit-html/src/test/directives/style-map_test.ts:48]: Maybe a TODO to revisit this?

**Review by kevinpschaaf — APPROVED** [review]
LGTM, after you've looked at @jridgewell's feedback also.

File: packages/lit-html/src/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/class-map.ts:88]: nice!

File: packages/lit-html/src/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/class-map.ts:134]: That's a good point. Overall I suspect this isn't going to be the low-hanging fruit in IE11 perf anyway. I'm ok with requiring a polyfill.

File: packages/lit-html/src/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/class-map.ts:51]: removed ClassList

File: packages/lit-html/rollup.config.js
Comment (justinfagnani) [reply packages/lit-html/rollup.config.js:49]: will do soon

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:770]: We've been preferring `Array<T>` recently to be consistent with more complex types. I'll confer with the team.

File: packages/lit-html/src/lit-html.ts
Comment (justinfagnani) [reply packages/lit-html/src/lit-html.ts:640]: indeed

File: packages/lit-html/src/directives/style-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/style-map.ts:72]: It looks like there a WebKit regex bug preventing this from working. `/(^a)?B/` matches "aB" but not "bB"

File: packages/lit-html/src/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/class-map.ts:96]: That happens in AttributePart now. We return a string with the dynamic class and that's interpolated into the static classes.

File: packages/lit-html/src/test/directives/style-map_test.ts
Comment (justinfagnani) [reply packages/lit-html/src/test/directives/style-map_test.ts:48]: I think this is ok actually. It's either this or exporting the class.

File: packages/lit-html/src/lib/directives/class-map.ts
Comment (sorvell) [reply packages/lit-html/src/lib/directives/class-map.ts]: Constructor now gets PartInfo. Update is the only place you get access to dom.

File: packages/lit-html/src/directives/style-map.ts
Comment (jridgewell) [reply packages/lit-html/src/directives/style-map.ts]: Wow, didn't know that. Let's try `replace(/^(webkit|moz|ms|o)(?=[A-Z])|[A-Z]/g, '-$&')`, then.

Comment (jridgewell) [reply packages/lit-html/src/directives/style-map.ts]: Ooo, looks like `replace(/(?:^(webkit|moz|ms|o)|)(?=[A-Z])/g, '-$&')` does the trick

Comment (jridgewell) [reply packages/lit-html/src/directives/style-map.ts]: Opened https://bugs.webkit.org/show_bug.cgi?id=216671

File: packages/lit-html/src/directives/style-map.ts
Comment (sorvell) [reply packages/lit-html/src/directives/style-map.ts:81]: I think we shouldn't block on addressing this as we've discussed modifying this.

File: packages/lit-html/src/directives/style-map.ts
Comment (sorvell) [inline packages/lit-html/src/directives/style-map.ts:60]: We could elect to require users to use the dash case, but the ergonomics are a bit annoying since you need to do `{[`-webkit-appearance`]: condition}`.

**Review by sorvell — APPROVED** [review]

File: packages/lit-html/src/directives/class-map.ts
Comment (jridgewell) [reply packages/lit-html/src/directives/class-map.ts:62]: Doesn't this also apply to `styleMap`'s initial render, then?

File: packages/lit-html/src/directives/style-map.ts
Comment (jridgewell) [inline packages/lit-html/src/directives/style-map.ts:72]: continuing https://github.com/Polymer/lit-html/pull/1218#discussion_r489158205 so it doesn't get lost:
```suggestion
        prop = prop.replace(/(?:^(webkit|moz|ms|o)|)(?=[A-Z])/g, '-$&').toLowerCase();
```

File: packages/lit-html/src/directives/style-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/style-map.ts:58]: That's too big of a breaking change

File: packages/lit-html/src/directives/class-map.ts
Comment (justinfagnani) [reply packages/lit-html/src/directives/class-map.ts:62]: That's true! I removed the writing of static styles there.

**Pull request merged**
